RefinedForGithub extension cannot be enabled: Invalid call to permissions.request()

elijahsawyers opened the pull request:
#### 027a604092f7e3765e9f22f44e71cbd8c506bf90
<pre>
RefinedForGithub extension cannot be enabled: Invalid call to permissions.request()
<a href="https://bugs.webkit.org/show_bug.cgi?id=270918">https://bugs.webkit.org/show_bug.cgi?id=270918</a>
<a href="https://rdar.apple.com/123775789">rdar://123775789</a>

Reviewed by Timothy Hatcher.

When requesting permissions (permissions.request()), there is a check to make sure
that the method was called when processing a user gesture. As it stands, the check
doesn&apos;t handle events for actions, context menus, and commands. This patch makes
it such that WebCore::UserGestureIndicator::processingUserGesture() returns true
when handling those events.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIActionCocoa.mm:
(WebKit::WebExtensionContextProxy::dispatchActionClickedEvent):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPICommandsCocoa.mm:
(WebKit::WebExtensionContextProxy::dispatchCommandsCommandEvent):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIMenusCocoa.mm:
(WebKit::WebExtensionContextProxy::dispatchMenusClickedEvent):

Canonical link: <a href="https://commits.webkit.org/276065@main">https://commits.webkit.org/276065@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/66890d6ce48edcd08451d0e501d0b9695e8acfff

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43642 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22690 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46080 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46284 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39771 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45946 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26512 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20098 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36057 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44216 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19704 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37578 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Running apply-patch; Checked out pull request; Running run-api-tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17038 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17253 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38639 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1699 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39798 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38950 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47834 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18682 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15278 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42848 "Passed tests") | 
| | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41519 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; 1 api test failed or timed out; 5 api tests failed or timed out; Running compile-webkit-without-change") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20280 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5959 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19731 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

elijahsawyers commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPICommands.mm:
```diff
@@ -43,6 +43,8 @@
     @"description": @"Test Commands",
     @"version": @"1.0",
 
+    @"permissions": @[ @"webNavigation" ],
```
Does it really matter what permission I choose to put here, or can I just arbitrarily choose one for the sake of the test (like I did here)?

xeenon replied:
It can be anything that is valid really.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPICommands.mm:
```diff
@@ -319,6 +321,45 @@
+    EXPECT_EQ(filteredCommands.count, 1lu);
+
+    [manager.get().context performCommand:filteredCommands.firstObject];
+    [manager run];
```
Add a newline between.

elijahsawyers committed:
RefinedForGithub extension cannot be enabled: Invalid call to permissions.request()
https://bugs.webkit.org/show_bug.cgi?id=270918
rdar://123775789

Reviewed by Timothy Hatcher.

When requesting permissions (permissions.request()), there is a check to make sure
that the method was called when processing a user gesture. As it stands, the check
doesn't handle events for actions, context menus, and commands. This patch makes
it such that WebCore::UserGestureIndicator::processingUserGesture() returns true
when handling those events.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIActionCocoa.mm:
(WebKit::WebExtensionContextProxy::dispatchActionClickedEvent):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPICommandsCocoa.mm:
(WebKit::WebExtensionContextProxy::dispatchCommandsCommandEvent):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIMenusCocoa.mm:
(WebKit::WebExtensionContextProxy::dispatchMenusClickedEvent):

Canonical link: https://commits.webkit.org/276065@main
